File: corekit/src/main/java/com/laohu/kit/util/crash/CrashSaver.java
package com.laohu.kit.util.crash;

import android.content.Context;
import android.text.TextUtils;

import com.laohu.kit.util.MD5;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.LineNumberReader;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Iterator;
import java.util.Map;

class CrashSaver {

    public static void save(Context context, Throwable ex, boolean uncaught, Map<String, String> appendSnapshot) {
        Writer writer = null;
        PrintWriter printWriter = null;
        String stackTrace = "";
        try {
            writer = new StringWriter();
            printWriter = new PrintWriter(writer);
            ex.printStackTrace(printWriter);
            Throwable cause = ex.getCause();
            while (cause != null) {
                cause.printStackTrace(printWriter);
                cause = cause.getCause();
            }
            stackTrace = writer.toString();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (printWriter != null) {
                printWriter.close();
            }
        }
        String signature = stackTrace.replaceAll("\\([^\\(]*\\)", "");
        String filename = MD5.getStringMD5(signature);
        if (TextUtils.isEmpty(filename)) {
            return;
        }
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        Date date = new Date();
        String timestamp = sdf.format(date);
        BufferedWriter mBufferedWriter = null;
        try {
            File mFile = new File(context.getExternalFilesDir(null), "crash/" + filename + ".log");
            File pFile = mFile.getParentFile();
            if (!pFile.exists()) {// 如果文件夹不存在，则先创建文件夹
                pFile.mkdirs();
            }
            int count = 1;
            if (mFile.exists()) {
                LineNumberReader reader = null;
                try {
                    reader = new LineNumberReader(new FileReader(mFile));
                    String line = reader.readLine();
                    if (line != null && line.startsWith("count")) {
                        int index = line.indexOf(":");
                        if (index != -1) {
                            String count_str = line.substring(++index);
                            if (count_str != null) {
                                count_str = count_str.trim();
                                count = Integer.parseInt(count_str);
                                count++;
                            }
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                } finally {
                    if (reader != null) {
                        try {
                            reader.close();
                        } catch (Exception e) {
                            e.printStackTrace();
                        }

                    }
                }
                mFile.delete();
            }

            mFile.createNewFile();

            mBufferedWriter = new BufferedWriter(new FileWriter(mFile, true));// 追加模式写文件
            Map<String, String> snapshot = CrashSnapshot.snapshot(context, uncaught, timestamp, count);
            if (appendSnapshot != null) {
                snapshot.putAll(appendSnapshot);
            }
            String content = buildSnapshotString(snapshot, stackTrace);
            mBufferedWriter.append(content);
            mBufferedWriter.flush();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (mBufferedWriter != null) {
                try {
                    mBufferedWriter.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private static String buildSnapshotString(Map<String, String> snapshot, String trace) {
        Iterator<Map.Entry<String, String>> iterator = snapshot.entrySet().iterator();
        StringBuilder sb = new StringBuilder();
        while (iterator.hasNext()) {
            Map.Entry<String, String> entry = iterator.next();
            if (entry != null) {
                sb.append(entry.getKey()).append(entry.getValue());
                sb.append(System.getProperty("line.separator"));
            }
        }
        sb.append(System.getProperty("line.separator"));
        sb.append("**----------------------***");
        sb.append(System.getProperty("line.separator"));
        sb.append(trace);
        sb.append(System.getProperty("line.separator"));
        sb.append(System.getProperty("line.separator"));
        sb.append(System.getProperty("line.separator"));
        return sb.toString();
    }
}

File: corekit/src/main/java/com/laohu/kit/util/crash/CrashSaver.kt
package com.laohu.kit.util.crash

import android.content.Context
import android.text.TextUtils
import com.laohu.kit.extensions.orFalse
import com.laohu.kit.util.MD5
import java.io.BufferedWriter
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.io.LineNumberReader
import java.io.PrintWriter
import java.io.StringWriter
import java.io.Writer
import java.text.SimpleDateFormat
import java.util.*

object CrashSaver {
    fun save(
        context: Context,
        throwable: Throwable?,
        uncaught: Boolean,
        appendedSnapshot: Map<String, String>?
    ) {
        if (throwable == null) {
            return
        }
        var writer: Writer? = null
        var printWriter: PrintWriter? = null
        var stackTrace = ""
        try {
            writer = StringWriter()
            printWriter = PrintWriter(writer)
            throwable.printStackTrace(printWriter)
            var cause = throwable.cause
            while (cause != null) {
                cause.printStackTrace(printWriter)
                cause = cause.cause
            }
            stackTrace = writer.toString()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                writer?.close()
                printWriter?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        val signature = stackTrace.replace("\\([^\\(]*\\)".toRegex(), "")
        val filename = MD5.getStringMD5(signature)
        if (TextUtils.isEmpty(filename)) {
            return
        }
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val timestamp = sdf.format(Date())
        var mBufferedWriter: BufferedWriter? = null
        try {
            val file = File(context.getExternalFilesDir(null), "crash/$filename.log")
            if (!file.parentFile?.exists().orFalse()) { // 如果文件夹不存在，则先创建文件夹
                file.parentFile?.mkdirs()
            }
            var count = 1
            if (file.exists()) {
                var reader: LineNumberReader? = null
                try {
                    reader = LineNumberReader(FileReader(file))
                    val line = reader.readLine()
                    if (line != null && line.startsWith("count")) {
                        var index = line.indexOf(":")
                        if (index != -1) {
                            var countStr = line.substring(++index)
                            countStr = countStr.trim { it <= ' ' }
                            count = countStr.toInt()
                            count++
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                } finally {
                    if (reader != null) {
                        try {
                            reader.close()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                }
                file.delete()
            }
            file.createNewFile()
            mBufferedWriter = BufferedWriter(FileWriter(file, true)) // 追加模式写文件
            val snapshot: MutableMap<String, String> =
                CrashSnapshot.snapshot(context, uncaught, timestamp, count)
            if (appendedSnapshot != null) {
                snapshot.putAll(appendedSnapshot)
            }
            val content = buildSnapshotString(snapshot, stackTrace)
            mBufferedWriter.append(content)
            mBufferedWriter.flush()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                mBufferedWriter?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun buildSnapshotString(snapshot: Map<String, String>, trace: String): String {
        val iterator = snapshot.entries.iterator()
        val sb = StringBuilder()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            sb.append(entry.key).append(entry.value)
            sb.append(System.getProperty("line.separator"))
        }
        sb.append(System.getProperty("line.separator"))
        sb.append("**----------------------***")
        sb.append(System.getProperty("line.separator"))
        sb.append(trace)
        sb.append(System.getProperty("line.separator"))
        sb.append(System.getProperty("line.separator"))
        sb.append(System.getProperty("line.separator"))
        return sb.toString()
    }
}